### ISSUE TITLE
Better workflow for "Author is Owner" / "No Other Owners" 

### ISSUE THREAD

**Issue by Jojoshua** [DefinitelyTyped/dt-mergebot#105]
Since recently, the demand for updating single owner definitions outstrips the capacity for DT maintainers to perform reviews. It used to be that I could count on adding features to my typings and it be no longer than 2 or 3 days and it would be merged but it's changed since the bot has taken over more responsibility. I am looking for a better workflow for single owner definitions. 

The single owner scenario should be treated with responsibility of maintaining the quality of the package, not a DT maintainer. It takes me about 8 hours to bring a certain definition up to date with a new JS release. A DT maintainer is not going to put as much effort into the review as I, the single owner, did and it's half insulting to assume so.

If the single owner is good enough for reviewing and allowing commits from others then they should surely be good enough for their own changes. They should be treated as if they own that part of the repo. In fact maybe having a way to allow git submodules to pull in definitions wouldn't be a bad idea (it would speed up the DT CI build). 

Looking at the current DT board. There are ~300 recently merged PRs...none of which are tagged "No Other Owners". Those are all waiting in other boards.

My current options are...
- Add another "owner" simply to allow the bot to ping someone else to review/approve the changes. Who is qualified? Who cares? The NPM Credits section shows "These definitions were written by" which would be technically incorrect unless it starts showing these definitions were Written By: Owners , Reviewed By: Everyone that's ever approved changes on the file.

- Wait for some unknown period of time until a volunteer comes along (~3 weeks?)

- Go try to hunt down a DT maintainer, and request their review. (Tried it and did not work)

- Become a DT maintainer? Does that allow me to approve my own changes?

- Forget using DT and maintain my own private repo.

**Comment by orta** [comment]
I think it's worth us thinking about whether we can support allowing self merging when the module is under a particular level of popularity

That said. I'm not sure it's as bad as you're pitching, we added the label "No other owners" 2 days ago and since then [we've merged 13 PRs](https://github.com/DefinitelyTyped/DefinitelyTyped/pulls?q=is%3Apr+label%3A%22No+Other+Owners%22+is%3Aclosed) and the oldest open PR is 24 days old which is just waiting on the PR author to self-merge. "No other owners" implies it goes directly into our DT TODO list instantly.

**Comment by Jojoshua** [comment]
I think that would be a good first step. If a definition ever reaches that popularity threshold then the likelihood of having more than a single owner goes up anyway.

**Comment by elibarzilay** [comment]
I improved things in several ways:
* First, a maintainer can "bless" a PR, which is used to dismiss the need for a maintainer review.  It's supposed to serve as an indication of "I had a brief look and it looks ok, but I'm not an expert" on the maintainer side, which should make turnaround times faster.
* In the case of single owners of packages that are popular enough to require a maintainer, I'm making it
  - accept an "other" review even if it's popular enough to require a second owner review
  - spit out a message to the author suggesting finding someone else to approve it and maybe even be added to the owners

Re that last thing, I think that it's a perfectly valid way to speed things up --- for cases when a single owner can find someone else, things are better in the sense that there's a second pair of eyes that looked at the code, and adding them as an owner should improve this in future PRs too.  Of course you can game the system and make up another GH account or ask your neighbor, but that's an effort to save for the few days it'll take for the PR to get to a maintainer anyway.

 (And yes, maybe there should be another category in the future for "soft owners" who don't contribute code, I'm guessing that in such a situation people would likely add some comment anyway.)

**Comment by Jojoshua** [comment]
> I think it's worth us thinking about whether we can support allowing self merging when the module is under a particular level of popularity
> 
> That said. I'm not sure it's as bad as you're pitching, we added the label "No other owners" 2 days ago and since then [we've merged 13 PRs](https://github.com/DefinitelyTyped/DefinitelyTyped/pulls?q=is%3Apr+label%3A%22No+Other+Owners%22+is%3Aclosed) and the oldest open PR is 24 days old which is just waiting on the PR author to self-merge. "No other owners" implies it goes directly into our DT TODO list instantly.

Could this idea of popularity be revisited @orta ? I have not observed this PR really addressed the original issue now with another PR 2 weeks out https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49737. The bot can beg for reviews but that isn't helping.

**Comment by elibarzilay** [comment]
Good morning,

You're right -- that PR should be on the maintainer queue.  This is a
bug in additional changes that were added last week: that PR should not
have gone into the "Unreviewed" timeline because there are no other
owners.  (The ping for "«anyone?»" there is an indication that something
unexpected happened.)  With #254, the PR would be on the maintainers
queue.  And yes, before these changes there were other problems with the
timelines, which is why it ended up for two weeks and didn't get to the
maintainer queue.

Re your original issue: things have changed a lot since then, but
handling PRs with a single owner is still similar overall: either
there's a passing review (and based on popularity as a rough scale for
importance, you can get a review from anyone).  You're right that this
is a potential bottleneck, and you're right that a DT maintainer cannot
compete with the amount of time that an author put in creating a PR.
But the idea is to have *any* second set of eyes overlook all commits,
which is a pretty acceptable requirement.

(Owners are, BTW, not the same as the people who wrote the code --- they
are rather the people who care about a typing package enough to review
PRs.  Some of the changes since the summer are an attempt to clarify
it.)

----

And a final footnote: there are currently 8 PRs with "No Other Owners":

* 3 waiting for author (2 for a self merge, 1 for changereqs)
* 1 on the maintainer queue
* 4 (including #49737) are in waiting-for-review, and this PR will move
  them to the queue.

And this was a relatively difficult week for the maintainer, since the
changes last week made a flood of PRs get onto the queue.

It might be a good idea to have a quick version of the unreviewed
timeline, which is quicker than the normal one, in case owners can find
someone to review a PR, but it looks like this is not a high load to be
a problem in practice.

**Comment by Jojoshua** [comment]
> But the idea is to have _any_ second set of eyes overlook all commits,
> which is a pretty acceptable requirement.

In general software development I would agree with having a 2nd set of eyes but TypeScript definitions seem different and especially this particular scenario. 

Can the ts-bot count as a 2nd set of eyes for this case...I mean it did provide us with a warm fuzzy... 
>It looks like nothing changed too much.

Perhaps my knowledge of the overall TS ecosystem is limited but I am asking myself... 

_"What's the worst that can happen?"_
- Malicious activity? -- When was the last time a TS definition caused a security vulnerability?
- Fat finger? -- If there was an accidental problem discovered, who is going to fix it? A DT maintainer or the authors?"

_"How often are DT maintainer comments on reviews are **substantive** vs rubber stamping?"_

- If an issue was pushed out, the consuming dev team can easily rollback to the previous version of the definition while it get fixed. 
- If there is a problematic author, why not treat that as an **exception** and introduce a way to vote for good maintainers, have the ts-bot give points for each good PR(with the goal of eventually be automatic merge) or remove points for bad PR, or have a way to ban bad authors until they remediate.

**Comment by RyanCavanaugh** [comment]
@Jojoshua we hear your feedback; the PR delays you experienced recently were the result of a delay, and you shouldn't encounter these again in the future.

We have three main objectives here, in this order:
 * Don't damage the `@types` scope in any way
 * Encourage community to be self-managing and feel a sense of ownership
 * Turn around PRs as quickly as is reasonable

The worst things that can happen are honestly pretty bad. Editors use contents from DefinitelyTyped to inform the editor experience, e.g. offensive information being merged in is bad. Type definitions can (and have) cause major responsiveness slowdowns if they're written too compexly. If someone accidently publishes a major version number that's too high, we have to personally engage with npm support to undo that, and on a short timescale. Sometimes people don't even realize the implications of changes they're making and maintainers need to confirm that they understand what's going to happen if they merge a PR.

Overall this is one of the busiest repos on GitHub and is staffed by one individual working a normal work week; we are comfortable (after bug fixes) with how the current workflow balances safety vs expedience given those constraints. We'll continue to assess this, and rest assured that our incentives are entirely aligned with yours here --- no one really enjoys looking at several hundred PRs a week, so anything we think is safe to automate away, we already have.

**Comment by Jojoshua** [comment]
> @Jojoshua we hear your feedback; the PR delays you experienced recently were the result of a bug, and you shouldn't encounter these again in the future.
> 
> We have three main objectives here, in this order:
> 
> * Don't damage the `@types` scope in any way
> * Encourage community to be self-managing and feel a sense of ownership
> * Turn around PRs as quickly as is reasonable
> 
> The worst things that can happen are honestly pretty bad. Editors use contents from DefinitelyTyped to inform the editor experience, e.g. offensive information being merged in is bad. Type definitions can (and have) cause major responsiveness slowdowns if they're written too compexly. If someone accidently publishes a major version number that's too high, we have to personally engage with npm support to undo that, and on a short timescale. Sometimes people don't even realize the implications of changes they're making and maintainers need to confirm that they understand what's going to happen if they merge a PR.
> 
> Overall this is one of the busiest repos on GitHub and is staffed by one individual working a normal work week; we are comfortable (after bug fixes) with how the current workflow balances safety vs expedience given those constraints. We'll continue to assess this, and rest assured that our incentives are entirely aligned with yours here --- no one really enjoys looking at several hundred PRs a week, so anything we think is safe to automate away, we will.

@RyanCavanaugh , thank you for the information and thoughtful response. I also appreciate the quick fix @elibarzilay . Will the fix auto-heal my PR?
